test/unittests/evm_test.cpp

```diff
@@ -78,6 +78,154 @@ TEST_F(evm, sub_and_swap)
+
+    execute(pushes + OP_DUPN + "0000" + ret_top());
+    EXPECT_STATUS(EVMC_SUCCESS);
+    EXPECT_OUTPUT_INT(4);
+
+    execute(pushes + OP_DUPN + "0002" + ret_top());
+    EXPECT_STATUS(EVMC_SUCCESS);
+    EXPECT_OUTPUT_INT(2);
+
+    execute(pushes + OP_DUPN + "0003" + ret_top());
+    EXPECT_STATUS(EVMC_SUCCESS);
+    EXPECT_OUTPUT_INT(1);
+
+    execute(pushes + OP_DUPN + "0004" + ret_top());
+    EXPECT_STATUS(EVMC_STACK_UNDERFLOW);
+
+    execute(pushes + OP_DUPSN + "00" + ret_top());
+    EXPECT_STATUS(EVMC_SUCCESS);
+    EXPECT_OUTPUT_INT(4);
+
+    execute(pushes + OP_DUPSN + "01" + ret_top());
+    EXPECT_STATUS(EVMC_SUCCESS);
+    EXPECT_OUTPUT_INT(3);
+
+    execute(pushes + OP_DUPSN + "03" + ret_top());
+    EXPECT_STATUS(EVMC_SUCCESS);
+    EXPECT_OUTPUT_INT(1);
+
+    execute(pushes + OP_DUPSN + "04" + ret_top());
+    EXPECT_STATUS(EVMC_STACK_UNDERFLOW);
+}
+
+TEST_F(evm, swapn)
+{
+    rev = EVMC_ISTANBUL;
+    const auto pushes = push(1) + push(2) + push(3) + push(4);
+
+    execute(pushes + OP_SWAPN + "0000" + ret_top());
+    EXPECT_STATUS(EVMC_SUCCESS);
+    EXPECT_OUTPUT_INT(3);
+
+    execute(pushes + OP_SWAPN + "0001" + ret_top());
+    EXPECT_STATUS(EVMC_SUCCESS);
+    EXPECT_OUTPUT_INT(2);
+
+    execute(pushes + OP_SWAPN + "0002" + ret_top());
+    EXPECT_STATUS(EVMC_SUCCESS);
+    EXPECT_OUTPUT_INT(1);
+
+    execute(pushes + OP_SWAPN + "0003" + ret_top());
+    EXPECT_STATUS(EVMC_STACK_UNDERFLOW);
+
+    execute(pushes + OP_SWAPSN + "00" + ret_top());
+    EXPECT_STATUS(EVMC_SUCCESS);
+    EXPECT_OUTPUT_INT(3);
+
+    execute(pushes + OP_SWAPSN + "01" + ret_top());
+    EXPECT_STATUS(EVMC_SUCCESS);
+    EXPECT_OUTPUT_INT(2);
+
+    execute(pushes + OP_SWAPSN + "02" + ret_top());
+    EXPECT_STATUS(EVMC_SUCCESS);
+    EXPECT_OUTPUT_INT(1);
+
+    execute(pushes + OP_SWAPSN + "03" + ret_top());
+    EXPECT_STATUS(EVMC_STACK_UNDERFLOW);
+}
+
+TEST_F(evm, dupn_full_stack)
+{
+    rev = EVMC_ISTANBUL;
+    auto full_stack_code = bytecode{};
+    for (int i = 1023; i >= 1; --i)
+        full_stack_code += push(i);
+
+    execute(full_stack_code + OP_POP + OP_DUPSN + "00" + ret_top());
+    EXPECT_STATUS(EVMC_SUCCESS);
+    EXPECT_OUTPUT_INT(2);
+
+    execute(full_stack_code + OP_POP + OP_DUPSN + "ff" + ret_top());
+    EXPECT_STATUS(EVMC_SUCCESS);
+    EXPECT_OUTPUT_INT(255 + 2);
+
+    execute(full_stack_code + OP_DUPSN + "01" + OP_DUPSN + "02");
+    EXPECT_STATUS(EVMC_STACK_OVERFLOW);
+
+    execute(full_stack_code + OP_POP + OP_DUPN + "0000" + ret_top());
+    EXPECT_STATUS(EVMC_SUCCESS);
+    EXPECT_OUTPUT_INT(2);
+
+    execute(full_stack_code + OP_POP + OP_DUPN + "0100" + ret_top());
+    EXPECT_STATUS(EVMC_SUCCESS);
+    EXPECT_OUTPUT_INT(256 + 2);
+
+    execute(full_stack_code + OP_POP + OP_DUPN + "03fd" + ret_top());
+    EXPECT_STATUS(EVMC_SUCCESS);
+    EXPECT_OUTPUT_INT(1023);
+
+    execute(full_stack_code + OP_DUPN + "03fe" + OP_SWAPSN + "00" + OP_RETURN);
```

Comment:
why does it not use `ret_top()` and how does it return anything without storing in memory?

Reply:
The `ret_top()` requires additional push therefore more stack space. We only check the output size which matches the last duped item.